src/swtpm/ctrlchannel.c

```diff
@@ -220,6 +222,157 @@ static int ctrlchannel_receive_state(ptm_setstate *pss, ssize_t n, int fd)
     return fd;
 }
 
+struct input {
+    uint32_t cmd;
+    /* ptm_hdata is the largest buffer to receive */
+    uint8_t body[sizeof(ptm_hdata)];
+} input;
+
+/* timespec_diff: calculate difference between two timespecs
+ *
+ * @end: end time
+ * @start: start time; must be earlier than @end
+ * @diff: result
+ *
+ * This function will return a negative tv_sec in result, if
+ * @end is earlier than @start, the time difference otherwise.
+ */
+static void timespec_diff(struct timespec *end,
+                          struct timespec *start,
+                          struct timespec *diff)
+{
+    diff->tv_nsec = end->tv_nsec - start->tv_nsec;
+    diff->tv_sec = end->tv_sec - start->tv_sec;
+    if (diff->tv_nsec < 0) {
+        diff->tv_nsec += 1E9;
+        diff->tv_sec -= 1;
+    }
+}
+
+/*
+ * ctrlchannel_recv_cmd: Receive a command on the control channel
+ *
+ * @fd: file descriptor for control channel
+ * @buffer: buffer to receive data into
+ * @buffer_len: length of the buffer
+ *
+ * This function returns 0 or a negative number if an error receiving
+ * the command occurred, including a timeout. In case of success,
+ * the nunber of bytes received is returned.
+ */
+static ssize_t ctrlchannel_recv_cmd(int fd,
+                                    unsigned char *buffer,
+                                    size_t buffer_len)
+{
+    ssize_t n;
+    size_t recvd = 0;
+    size_t needed = offsetof(struct input, body);
+    struct input *input = (struct input *)buffer;
+    struct pollfd pollfd =  {
+        .fd = fd,
+        .events = POLLIN,
+    };
+    struct timespec deadline, now, timeout;
+    int to;
+    /* Read-write */
+    ptm_init *init_p;
+    ptm_reset_est *pre;
+    ptm_hdata *phd;
+    ptm_getstate *pgs;
+    ptm_setstate *pss;
+    ptm_loc *pl;
+
+    clock_gettime(CLOCK_REALTIME, &deadline);
+
+    /* maximum allowed time is 500ms to recv everything */
+    deadline.tv_nsec += 500 * 1E6;
+    if (deadline.tv_nsec >= 1E9) {
+        deadline.tv_nsec -= 1E9;
+        deadline.tv_sec += 1;
+    }
+
+    while (recvd < buffer_len) {
+        n = read(fd, &buffer[recvd], buffer_len - recvd);
+        if (n <= 0)
+            return n;
+        recvd += n;
+        /* we need to at least see the cmd */
+        if (recvd < offsetof(struct input, body))
+            goto wait_chunk;
```

Comment:
Do we really need to wait here, I mean do we need to support receiving ctrl 'command' also in chunks ?, cant we treat it as some went wrong and return.

Reply:
It's really only these two additional lines that allow us to even wait for that...